stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -47,6 +47,11 @@ julia> Diagonal(V)
 """
 Diagonal(V::AbstractVector)
 
+Diagonal(diag::V) where {S,V<:AbstractVector{S}} = Diagonal{promote_type(S, typeof(zero(S))),V}(diag)
```

Comment:
I don't think we want promotion here.  I think `Union{S, typeof(zero(S))}` would be better.

Reply:
What is your reasoning for this?

Reply:
Promotion isn't necessarily a widening operation:

```
julia> promote_type(Int, Float64)
Float64
```

It's often used alongside conversions to combine multiple arguments into a single result.  Here, we're not doing any combining or converting — it's just saying we'll either return something of type `S` or `zero(S)`.

Now, we could use `typejoin`, which will choose the best type or supertype that encompasses its arguments, but I think that's overly pessimistic.

Reply:
Do note that as it stands, I'm explicitly calling the `T` constructor in `getindex` in this PR, to make sure that what's returned from `getindex` is actually of type `T` (perhaps this should be a `convert`, not 100% up to date with the changes on that front). So in your example, if `zero(::Type{Int})` were a `Float64`, conversion to `Float64` would happen for on-diagonal elements in `getindex`, which I think is exactly the right thing.